### PR TITLE
Add warning to suggest to use inline classes when a class has single property

### LINKED ISSUE #698: Add warning to suggest to use inline classes when a class has single property
https://kotlinlang.org/docs/reference/inline-classes.html

However, only since 1.3. Maybe needs some more considerations.

### PR DESCRIPTION
**Pull request by aktsay6**
### What's done:
  * Logic made
  * Added warn tests

This pull request closes #698 

## Actions checklist
* [X] Implemented Rule, added Warnings
* [X] Added tests on checks
* [ ] Added tests on fixers
* [X] Updated diktat-analysis.yml
* [X] Updated available-rules.md

## Fixme
Since it is still experimental feature it shouldn't be fixed

### CODE REVIEW

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/classes/InlineClassesRule.kt [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+        if (node.elementType == CLASS) {
+            handleClasses(node)
+        }
+    }
+
+    private fun handleClasses(node: ASTNode) {
+        if ((node.psi as KtClass).getProperties().size == 1 && !(node.psi as KtClass).hasExplicitPrimaryConstructor()) {
+            val modList = node.getFirstChildWithType(MODIFIER_LIST)
+            if (modList == null || modList.getChildren(null).all { it.elementType in goodModifiers }) {
```
Comment (kentr0w) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/classes/InlineClassesRule.kt:39]: Why shouldn't be constructor?

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/classes/InlineClassesRule.kt [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        emitWarn = emit
+        isFixMode = autoCorrect
+
+        if (node.elementType == CLASS) {
+            handleClasses(node)
+        }
+    }
+
+    private fun handleClasses(node: ASTNode) {
+        if ((node.psi as KtClass).getProperties().size == 1 && !(node.psi as KtClass).hasExplicitPrimaryConstructor()) {
```
Comment (kentr0w) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/classes/InlineClassesRule.kt:37]: If there is `var`?

**Review by kentr0w — CHANGES_REQUESTED** [review]

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/classes/InlineClassesRule.kt [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+            val modList = node.getFirstChildWithType(MODIFIER_LIST)
```
Comment (kentr0w) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/classes/InlineClassesRule.kt:38]: Here you can also use `node.psi as KtClass` to get all modifiers and pass to this method node as psi

**Review by kentr0w — CHANGES_REQUESTED** [review]

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/classes/InlineClassesRule.kt (hunk 1 above)
Comment (kentr0w) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/classes/InlineClassesRule.kt:39]: You can simplify this expression

**Review by kentr0w — CHANGES_REQUESTED** [review]

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/classes/InlineClassesRule.kt (hunk 1 above)
Comment (aktsay6) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/classes/InlineClassesRule.kt]: Fixed

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/classes/InlineClassesRule.kt (hunk 2 above)
Comment (aktsay6) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/classes/InlineClassesRule.kt]: Fixed

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/classes/InlineClassesRule.kt (hunk 3 above)
Comment (aktsay6) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/classes/InlineClassesRule.kt]: Fixed

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/classes/InlineClassesRule.kt (hunk 1 above)
Comment (aktsay6) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/classes/InlineClassesRule.kt]: Fixed

File: diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/chapter6/InlineClassesWarnTest.kt [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+                |class Some : Any {
+                |   val some = 3
+                |}
+            """.trimMargin()
+        )
+    }
+
+    @Test
+    @Tag(WarningNames.INLINE_CLASS_CAN_BE_USED)
+    fun `should not trigger on class with internal constructor`() {
```
Comment (petertrr) [inline diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/chapter6/InlineClassesWarnTest.kt:117]: Are you sure it's prohibited? IDEA doesn't seem to highlight it in red.

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/classes/InlineClassesRule.kt [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        emitWarn = emit
+        isFixMode = autoCorrect
+
+        if (node.elementType == CLASS) {
+            handleClasses(node.psi as KtClass)
+        }
+    }
+
+    private fun handleClasses(classPsi: KtClass) {
+        // Fixme: for now we can't understand whether it extends class or interface
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/classes/InlineClassesRule.kt:43]: If it extends class, it's constructor should be called, even if it doesn't accept parameters: `B : A()` vs `B : I`. Or am I missing something?

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/classes/InlineClassesRule.kt [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+                classPsi.node.getFirstChildWithType(MODIFIER_LIST)?.getChildren(null)?.all { it.elementType in goodModifiers } != false) {
+            INLINE_CLASS_CAN_BE_USED.warnAndFix(configRule, emitWarn, isFixMode, "class ${classPsi.name}", classPsi.node.startOffset, classPsi.node) {
+                // Fixme: since it's an experimental feature we shouldn't do fixer
+            }
+        }
+    }
+
+    private fun hasValidProperties(classPsi: KtClass): Boolean {
+        if (classPsi.getProperties().size == 1 && !classPsi.hasExplicitPrimaryConstructor()) {
+            return !classPsi.getProperties().first().isVar
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/classes/InlineClassesRule.kt:55]:
```suggestion
            return !classPsi.getProperties().single().isVar
```
imho, using `single` better represents your intention in such cases

**Review by petertrr — CHANGES_REQUESTED** [review]

File: info/guide/guide-chapter-6.md [hunk 7]
```diff
@@ -357,6 +357,9 @@ fun main() {
 }
 ```
 
+### <a name="c6.1.12"></a> 6.1.12 Prefer Inline classes when a class has a single property
```
Comment (petertrr) [inline info/guide/guide-chapter-6.md:360]: Also please add it to TOC

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/classes/InlineClassesRule.kt (hunk 5 above)
Comment (aktsay6) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/classes/InlineClassesRule.kt]: Yeah, fixed

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/classes/InlineClassesRule.kt (hunk 6 above)
Comment (aktsay6) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/classes/InlineClassesRule.kt]: Fixed

File: info/guide/guide-chapter-6.md (hunk 7 above)
Comment (aktsay6) [reply info/guide/guide-chapter-6.md]: Added

File: diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/chapter6/InlineClassesWarnTest.kt (hunk 4 above)
Comment (petertrr) [reply diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/chapter6/InlineClassesWarnTest.kt:128]: idk, my idea doesn't :)
image

Comment (aktsay6) [reply diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/chapter6/InlineClassesWarnTest.kt:128]: Strange, I have this 
image
IntelliJ IDEA 2020.3 (Ultimate Edition)
Build #IU-203.5981.155, built on December 1, 2020

Comment (petertrr) [reply diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/chapter6/InlineClassesWarnTest.kt:128]: hmm, do you try this code in diktat? I would expect results to be the same with same kotlin version. Which version of kotlin plugin do you have in idea? My is 203-1.4.30-RC-IJ6682.9 for idea 2020.3.2. Let's check if this code compiles with kotlin 1.4.21.

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/classes/InlineClassesRule.kt [hunk 8]
```diff
@@ -52,15 +54,19 @@ class InlineClassesRule(private val configRule: List<RulesConfig>) : Rule("inlin
 
     private fun hasValidProperties(classPsi: KtClass): Boolean {
         if (classPsi.getProperties().size == 1 && !classPsi.hasExplicitPrimaryConstructor()) {
-            return !classPsi.getProperties().first().isVar
+            return !classPsi.getProperties().single().isVar
         } else if (classPsi.getProperties().isEmpty() && classPsi.hasExplicitPrimaryConstructor()) {
             return classPsi.primaryConstructorParameters.size == 1 &&
                     !classPsi.primaryConstructorParameters.first().node.hasChildOfType(VAR_KEYWORD) &&
-                    classPsi.primaryConstructorModifierList == null
+                    classPsi.primaryConstructor?.visibilityModifierType()?.value ?: "public" == "public"
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/classes/InlineClassesRule.kt:61]:
```suggestion
                    classPsi.primaryConstructor?.visibilityModifierType()?.value?.let { it == "public" } ?: true
```
Maybe, something like this?

File: diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/chapter6/InlineClassesWarnTest.kt (hunk 4 above)
Comment (aktsay6) [reply diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/chapter6/InlineClassesWarnTest.kt:128]: Now I have 1.4.21 and this code doesn't compile

Comment (petertrr) [reply diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/chapter6/InlineClassesWarnTest.kt:128]: OK, that is a bug in IDEA plugin. Maybe, it'll be allowed in 1.4.30, but let's allow only public for now.

**Review by akuleshov7 — CHANGES_REQUESTED** [review]
go go go

**Review by petertrr — APPROVED** [review]

**Review by kentr0w — APPROVED** [review]
lgfm

File: info/guide/guide-chapter-6.md [hunk 9]
```diff
@@ -357,6 +357,9 @@ fun main() {
 }
 ```
 
+### <a name="r6.1.12"></a> 6.1.12 Prefer Inline classes when a class has a single property
```
Comment (akuleshov7) [inline info/guide/guide-chapter-6.md:360]: example?

File: diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/utils/RulesConfigYamlTest.kt [hunk 10]
```diff
@@ -17,6 +17,7 @@ import kotlinx.serialization.encodeToString
  * Special test that checks that developer has not forgotten to add his warning to a diktat-analysis.yml
  * This file is needed to be in tact with latest changes in Warnings.kt
  */
+@Suppress("INLINE_CLASS_CAN_BE_USED")
```
Comment (akuleshov7) [inline diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/utils/RulesConfigYamlTest.kt:20]: may be inline this Test?

**Review by akuleshov7 — CHANGES_REQUESTED** [review]
please look at comments

**Review by akuleshov7 — APPROVED** [review]
lgtm

**Pull request merged**